### DISCUSSION TITLE
[!] Google Api Error: badRequest: Only releases with status draft may be created on draft app. - Only releases with status draft may be created on draft app.

### DISCUSSION THREAD

**Discussion by aprilmintacpineda** [fastlane/fastlane#18293] [Q&A] [answered]
I'm trying to publish to PlayStore Internal track.

If I do:

```ruby
gradle(task: "clean bundleRelease")

upload_to_play_store(
  package_name: "com.aprmp.fastlane_deployment",
  track: "internal",
  aab: "app/build/outputs/bundle/release/app-release.aab",
  release_status: "draft"
)
```

It works, it ends sucessfully, but I have to go to Play Console and then manually roll out the release, if I do:

```ruby
gradle(task: "clean bundleRelease")

upload_to_play_store(
  package_name: "com.aprmp.fastlane_deployment",
  track: "internal",
  aab: "app/build/outputs/bundle/release/app-release.aab"
)
```

It will end with error

```
[!] Google Api Error: badRequest: Only releases with status draft may be created on draft app. - Only releases with status draft may be created on draft app.
```

Yes, my App is currently in "draft" mode, but I can manually release to internal tracks, so I assume I should be able to do that in the API as well, right?

**Comment by Thangaraj-M** [comment, marked as answer]
The first version of app should be released manually through the play console. Then only we can use the google API to release the next version from Fastlane.
If we try to release the first version of app from Fastlane, this error will come.

**Reply by aprilmintacpineda** [reply]
So, I must release production first?

**Reply by Thangaraj-M** [reply]
Not required to release in production. It can be released to alpha or beta track also.

**Reply by benomatis** [reply]
@Thangaraj-M I have the same issue... So if I set up an Alpha (closed) testing track in the Play Store, would the following set up work and approrpriately release the app for internal testing without having to go to Play Store afterwards?
```ruby
supply(
  track: "internal",
  track_promote_to: "internal",
)
```
Would I need to use `track_promote_to` at all...? Or would I need to use `rollout: '1'` to achieve this?

**Reply by Thangaraj-M** [reply]
@benomatis
Once the first version released in the Alpha track manually, then the next version can be released with supply from Fastlane.
To release in internal testing track, no need to set rollout and track_promote_to options. It will work with the below options.

supply(
  track: "internal",
  package_name: "<packange_name>",
  apk: <apk_path>
)

**Reply by benomatis** [reply]
I can confirm this worked for me, hope @aprilmintacpineda (the OP) can make it work as well.

**Reply by aprilmintacpineda** [reply]
Thank you! :bow:

**Reply by syed9569a** [reply]
I am still confused on this. What I have is an app that's in draft mode and I want to simply automate the release for my app on the internal track. I am utilising upload_to_play_store:

upload_to_play_store(
            track: 'internal',
            package_name: < package_name>,
            version_name: < version_name>,
            version_code:< version_code>,
            release_status: 'draft',
            skip_upload_metadata: true,
            skip_upload_images: true,
            skip_upload_screenshots: true,
)

This uploads but never releases the app. There is no alpha, beta or production set up, only internal track is set up and the app is in draft.

**Reply by benomatis** [reply]
@syed9569a as said above, do create an Alpha track at least, and that will then allow you to release. You can deactivate the track later and continue to release to the internal track. You will never be able to completely delete the Alpha track, but that should be fine.

**Reply by daraul** [reply]
How come this is a problem, though? Is it an API limitation?

**Reply by mxcihak** [reply]
So if I have a staging app which will never go to production and is used only for testing, I still need to fill all the app info, store info and screenshots and images just so I can upload to internal testing is that right?

**Reply by aprilmintacpineda** [reply]
You don't need to actually, if you only want internal testing.

**Reply by justincbeck** [reply]
Every time I have to go and start working on a new app and get it setup in Google, I want to shoot myself in the face. Why is it so convoluted. It's a total mess. I'll try what you say.

**Reply by toplac** [reply]
I can confirm manually generating a alpha release solves the issue.

**Reply by dataviruset** [reply]
So in order to make automatic releases through the API to the internal testing track, I must have an alpha release with app description, icons, screenshots, all the compliance etc? Sounds super annoying.

**Reply by toplac** [reply]
yes 😞 
but at least dummy values are ok (at least in our case)

**Reply by dataviruset** [reply]
Seems to work. But it seems that submitting for closed testing (alpha) will also submit the app for review. Wonder what's gonna happen if we get rejected now... :')

For reference, this is the command we use and it works.
`fastlane supply --package_name com.example.foo.bar --json_key_data "redacted" --track internal --aab foo-bar.aab`

Seems we didn't need to use `--track_promote_to internal` or `--release-status completed` explicitly.

**Reply by toplac** [reply]
we closed the alpha track afterwards and it is still working.

**Reply by PatriciaSauer** [reply]
Hi @toplac, did you create the alpha track and let it review or did you just create the alpha track and closed it right away? I'm really wondering how I can prepare everything before the first public release using fastlane. I tried uploading screenshots to my internal track but I didn't get it to work. So there seems to be a difference between internal and alpha track. Thanks in advance

**Reply by toplac** [reply]
Hello @PatriciaSauer, unfortunately i no longer remember and moved away from the project :)

**Reply by PatriciaSauer** [reply]
Too bad. Thank for answering anyways @toplac

**Comment by kschults** [comment]
Thanks for the info here.

Question: if I submit an Alpha build, will my app become visible on the app store?

**Reply by benomatis** [reply]
offtopic, but no, alpha is not public

**Reply by kschults** [reply]
Thanks for the quick reply. Just to be clear - my app is currently in "Draft" and not visible on the Play Store. If I submit an Alpha, that won't change, correct?

**Reply by benomatis** [reply]
only what you submit to production will be public

**Reply by kschults** [reply]
Right, I get that. I'm talking about the app as a whole, not a specific release. (I get that this is an edge case, only because I don't currently have _any_ public builds and the app as a whole is still in draft)

**Reply by benomatis** [reply]
this is not an edge case: whatever you publish to production (subject to approval) will be public, anything else won't

**Reply by kschults** [reply]
Okay, thank you for your help

**Comment by resistdesign** [comment]
What's the purpose of the draft?

If we need internal testers to test, the draft never reaches them.

Maybe the tool should just say "You can't really use this tool because your app is in draft mode".

It's hard to imagine the reason I would use a CLI tool in CI only to have to go and click something in the console anyway.

From what I understand, Google wants you to submit your app for review, even when you only want internal testing done on the release. Then you can remove your app from draft mode and maybe get releases to your testers via the API and such.

But it's not clear and maybe FastLane can be helpful where Google is not?

Any insight, on this, from anyone?

**Reply by iBotPeaches** [reply]
This is a limitation on the Google API to prevent abuse and fastlane cannot do much here. What happened a few years ago is malware applications were being created, using closed/internal tracks and spinning up extremely fast. I'm guessing from pure automation over the API. 

Back then you could API everything on a new app (send to review, etc) purely from an API. That changed and now requires you to go through one official Google review before you are allowed to use the API for uploads. That leads to those various errors you see here.

So I'm curious though from your perspective what you think fastlane could do here differently. I don't think changing error messages is a good idea - the raw error from Google is pretty self explanatory to me.

**Reply by resistdesign** [reply]
Yeah, I definitely understand that. I think maybe even just a warning or telling them they can't use it until they're out of draft mode. 

Or something a little bit more than just  making it feel like "hey, yes we uploaded it and you're good to go". 

The tiniest message could save people hours. 

And maybe even mostly because the error message about the draft makes you think you're supposed to upload a draft. 

But you're not. 

Do you see what I mean?

**Reply by tshedor** [reply]
@iBotPeaches @resistdesign I've got [a solution](https://github.com/fastlane/fastlane/pull/29999#issuecomment-4337980518) that could be viable when manually integrated. It could also be used to set the correct default `status` when uploading
